File: game/game-logic.ts
import { playMelodicSound, playSound } from "./audio"
import { Block } from "./block"
import { CellularAutomata } from "./cellular-automata"
import { Collectable } from "./collectable"
import { Crate } from "./crate"
import Entity from "./entity"
import { Food } from "./food"
import { checkLevelWon } from "./game"
import { entities, levelInfo, undoable } from "./game-state"
import { hitTestAllEntities, invertCollisionLayer, layersCollide, lineNoDiagonals, sortEntities, topLayer, translateEntity, withinLevel } from "./helpers"
import { Inverter } from "./inverter"
import { RectangularEntity } from "./rectangular-entity"
import Snake, { SnakeSegment } from "./snake"
import { CollisionLayer, Move, Tile } from "./types"

export function analyzeMoveAbsolute(snake: Snake, tile: Tile): Move {
  // Using Math.sign() here would lead to checking if moving to an adjacent tile is valid,
  // even when a further tile is in question.
  // const dirX = Math.sign(tile.x - snake.segments[0].x)
  // const dirY = Math.sign(tile.y - snake.segments[0].y)
  const deltaGridX = Math.round(tile.x - snake.segments[0].x)
  const deltaGridY = Math.round(tile.y - snake.segments[0].y)
  return analyzeMoveRelative(snake, deltaGridX, deltaGridY)
}
export function analyzeMoveRelative(snake: Snake, dirX: number, dirY: number): Move {
  const head = snake.segments[0]
  const deltaX = dirX * head.width
  const deltaY = dirY * head.height
  const x = head.x + deltaX
  const y = head.y + deltaY
  const hitsAhead = hitTestAllEntities(x, y, { ignoreTailOfSnake: snake.growOnNextMove ? undefined : snake })
  const hitsAllAlong = snake.segments.flatMap(segment => hitTestAllEntities(segment.x, segment.y))
  const encumbered = hitsAllAlong.some(hit =>
    hit.entity.solid &&
    hit.entity !== snake &&
    entities.indexOf(hit.entity) > entities.indexOf(snake) &&
    !(hit.entity instanceof Snake && snake.fusedSnakeIds.has(hit.entity.id))
  )

  // Prevent moving backwards when two segments long
  // (When one segment long, you can plausibly move in any direction,
  // and when more than two segments long, a body segment will be in the way,
  // but when two segments, normally the tail is excluded from hit testing,
  // so you would be allowed to double back, but it feels weird to do a 180° turn.)
  // This also prevents an overlapped snake from doubling back on itself,
  // by moving into a tile occupied by a snake which is on top of this snake.
  // (But that is just a special case of the rule that you shouldn't
  // be able to move into a tile occupied by a snake which is on top of this snake.)
  const movingBackwards =
    snake.segments.length > 1 &&
    dirX === Math.sign(snake.segments[1].x - head.x) &&
    dirY === Math.sign(snake.segments[1].y - head.y)

  // I think I will need to move to a system where the move is simulated and then checked for validity,
  // to avoid the complexity of adding exceptions to game state access, when answering hypotheticals.
  // This could also help with animating undo/redo, which currently replaces all the entities, resetting animation timers,
  // and for the level editor, where I'd like to check for collisions to show warnings,
  // (but allow the collisions to happen so that editing isn't its own puzzle.)
  // If moves are analyzed by checking for collisions within a whole game board,
  // it could share some code. Theoretically.

  // Push objects
  const entitiesToPush: Entity[] = []
  {
    const hit = hitsAhead.find(hit => hit.entity.solid)
    // TODO: try pushing other snakes too (or save that for when splitting snakes; only push the tail / "dead" half)
    // TODO: recursively push crates
    if (hit?.entity instanceof Crate) {
      // Check if the crate can be pushed
      const newTile = { x: hit.entity.x + deltaX, y: hit.entity.y + deltaY, width: hit.entity.width, height: hit.entity.height }
      const hitsAheadCrate = hitTestAllEntities(newTile.x, newTile.y, { ignoreTailOfSnake: snake })
      if (
        withinLevel(newTile) &&
        layersCollide(hit.entity.layer, head.layer) &&
        !layersCollide(topLayer(hitsAheadCrate), hit.entity.layer)
      ) {
        entitiesToPush.push(hit.entity)
        const boxedCollectable = hitsAhead.find(hit => hit.entity instanceof Collectable)
        if (boxedCollectable) {
          entitiesToPush.push(boxedCollectable.entity)
        }
      }
    }
  }
  // Ignore pushed objects as obstacles
  for (const entity of entitiesToPush) {
    const index = hitsAhead.findIndex(hit => hit.entity === entity)
    if (index !== -1) {
      hitsAhead.splice(index, 1)
    }
  }

  return {
    // snakeId: snake.id,
    snake,
    valid:
      (dirX === 0 || dirY === 0) &&
      (Math.abs(dirX) === 1 || Math.abs(dirY) === 1) &&
      withinLevel({ x, y, width: head.width, height: head.height }) &&
      !movingBackwards &&
      !encumbered &&
      !layersCollide(topLayer(hitsAhead), head.layer),
    encumbered,
    to: { x, y, width: head.width, height: head.height },
    delta: { x: deltaX, y: deltaY },
    entitiesThere: hitsAhead.map(hit => hit.entity),
    entitiesToPush,
  }
}

export function canMove(snake: Snake): boolean {
  return analyzeMoveRelative(snake, 1, 0).valid ||
    analyzeMoveRelative(snake, 0, 1).valid ||
    analyzeMoveRelative(snake, -1, 0).valid ||
    analyzeMoveRelative(snake, 0, -1).valid
}
export function takeMove(move: Move): void {
  // const snake = entities.find(entity => entity instanceof Snake && entity.id === move.snakeId) as Snake
  const snake = move.snake
  undoable()
  playSound('move')
  if (snake.growOnNextMove) {
    growSnake(snake)
    snake.growOnNextMove = false
  }
  const head = snake.segments[0]
  for (let i = snake.segments.length - 1; i > 0; i--) {
    const segment = snake.segments[i]
    const prev = snake.segments[i - 1]
    segment.x = prev.x
    segment.y = prev.y
  }
  head.x = move.to.x
  head.y = move.to.y
  head.width = move.to.width
  head.height = move.to.height
  // Sort entities so the snake is on top of anything it's moving onto.
  // This handles the visual as well as making it so
  // you can't double back while inside an inverse snake.
  // Exclude non-solid collectables, since, if you don't eat them (because they're a different color),
  // they should stay visible.
  // (Collectables are automatically sorted on top at level design time.)
  const ontoIndices = move.entitiesThere.filter(e => e.solid).map(e => entities.indexOf(e))
  const maxIndex = Math.max(...ontoIndices)
  const thisIndex = entities.indexOf(snake)
  if (thisIndex < maxIndex) {
    // Add before removing so relevant indices
    // stay valid for both splice calls.
    entities.splice(maxIndex + 1, 0, snake)
    entities.splice(thisIndex, 1)
  }
  // Push objects
  if (move.entitiesToPush.length > 0) {
    playSound('pushCrate', { playbackRate: Math.random() * 0.1 + 0.95, volume: 0.3 })
  }
  for (const entity of move.entitiesToPush) {
    translateEntity(entity, move.delta.x, move.delta.y)
    entities.splice(entities.indexOf(entity), 1)
    entities.push(entity)
  }
  // Ensure collectables are on top of crates, so that you can scoop up collectables inside crates to push them around.
  if (move.entitiesToPush.length > 0) {
    sortEntities()
  }
  // Eat collectables
  for (const entity of move.entitiesThere) {
    if (
      entity instanceof Collectable &&
      layersCollide(entity.layer, head.layer) &&
      !move.entitiesToPush.includes(entity)
    ) {
      entities.splice(entities.indexOf(entity), 1)
      if (entity instanceof Food) {
        snake.growOnNextMove = true
        if (!checkLevelWon()) {
          playMelodicSound('eat', snake.getNextMelodyIndex())
        }
      } else if (entity instanceof Inverter) {
        invertSnake(snake)
      }
    }
  }
  // Pull fused snakes
  for (const fusedSnakeId of snake.fusedSnakeIds) {
    const fusedSnake = entities.find(entity => entity instanceof Snake && entity.id === fusedSnakeId) as Snake
    if (!fusedSnake) {
      // Invalid fusedSnakeIds will be cleaned up when serializing.
      // It's easier than trying to actively clear relationships when entities are deleted.
      continue
    }
    // Assuming snakes are fused at the tails.
    dragSnake(fusedSnake, fusedSnake.segments.length - 1, snake.segments[snake.segments.length - 1])
  }
  // Update cellular automata
  updateCellularAutomata()
}

// TODO: DRY, copied from function `drag` in level-editor.ts
function dragSnake(dragging: Snake, draggingSegmentIndex: number, to: Tile) {
  const draggingSegment = dragging.segments[draggingSegmentIndex]
  if (
    draggingSegment.x !== to.x ||
    draggingSegment.y !== to.y
  ) {
    // Avoids diagonals and segments longer than 1 tile
    const from = { x: draggingSegment.x, y: draggingSegment.y } // needs copy since it's mutated and lineNoDiagonals is a generator, so it computes lazily
    // Skip the first point, since it's the same as the segment's current position
    for (const point of [...lineNoDiagonals(from, to)].slice(1)) {
      for (let i = dragging.segments.length - 1; i > draggingSegmentIndex; i--) {
        lead(dragging.segments[i - 1], dragging.segments[i])
      }
      for (let i = 0; i < draggingSegmentIndex; i++) {
        lead(dragging.segments[i + 1], dragging.segments[i])
      }
      draggingSegment.x = point.x
      draggingSegment.y = point.y
      // if (Snake.DEBUG_SNAKE_DRAGGING) {
      //   draw()
      // }
    }
  }
}
// TODO: DRY, copied from function `lead` in level-editor.ts
function lead(leader: SnakeSegment, follower: SnakeSegment) {
  follower.x = leader.x
  follower.y = leader.y
  // Portals may be able to change the size of the snake in the future
  follower.width = leader.width
  follower.height = leader.height
  // Visualize dragging algorithm while paused in the debugger
  // if (Snake.DEBUG_SNAKE_DRAGGING) {
  //   draw()
  // }
}

function growSnake(snake: Snake): void {
  const tail = snake.segments[snake.segments.length - 1]
  // This only works because SnakeSegment is a flat object.
  const newTail = { ...tail }
  snake.segments.push(newTail)
}
function invertSnake(snake: Snake): void {
  // See also: `invert` function in level-editor.ts
  const handledEntities = new Set<Entity>()
  const handledPositions = new Set<`${number},${number}`>()

  function handleEntity(entity: Entity) {
    if (handledEntities.has(entity)) {
      return
    }
    handledEntities.add(entity)
    if (entity instanceof RectangularEntity) {
      entity.layer = invertCollisionLayer(entity.layer)
    } else if (entity instanceof Snake) {
      for (const segment of entity.segments) {
        handlePosition(segment.x, segment.y)
        segment.layer = invertCollisionLayer(segment.layer)
      }
    }
  }

  function handlePosition(x: number, y: number) {
    if (handledPositions.has(`${x},${y}`)) {
      return
    }
    handledPositions.add(`${x},${y}`)
    const hits = hitTestAllEntities(x, y)
    for (const hit of hits) {
      handleEntity(hit.entity)
    }
  }

  handleEntity(snake)

  for (const pos of handledPositions) {
    const [x, y] = pos.split(',').map(Number)
    const hits = hitTestAllEntities(x, y)
    if (!hits.some((hit) => hit.entity instanceof Block)) {
      // add a white block where there was implicit black
      const block = new Block()
      block.layer = CollisionLayer.White
      block.x = x
      block.y = y
      entities.unshift(block)
    }
  }
}

function updateCellularAutomata() {
  const occupiedTiles = new Map<string, CellularAutomata>()
  for (const entity of entities) {
    if (entity instanceof CellularAutomata) {
      const key = `${entity.x},${entity.y}`
      occupiedTiles.set(key, entity)
    }
  }
  const newOccupiedTiles = new Set<string>()
  for (let y = 0; y < levelInfo.height; y++) {
    for (let x = 0; x < levelInfo.width; x++) {
      const key = `${x},${y}`
      const neighbors = [
        // `${x - 1},${y - 1}`,
        `${x},${y - 1}`,
        // `${x + 1},${y - 1}`,
        `${x + 1},${y}`,
        // `${x + 1},${y + 1}`,
        `${x},${y + 1}`,
        // `${x - 1},${y + 1}`,
        `${x - 1},${y}`,
      ]
      // Real inefficient, computing the top collision layer for each neighbor, filtering to ignore the cellular automata itself
      // Could precompute a 2D array (or Map) of top collision layers. Might be able to do this totally differently.
      // Not sure exactly how the cellular automata should interact with everything.
      const neighborCount = neighbors.filter(n => occupiedTiles.get(n)?.layer === invertCollisionLayer(topLayer(hitTestAllEntities(x, y).filter((hit) => !(hit.entity instanceof CellularAutomata))))).length

      // Conway's Game of Life rules (requires 8 neighbors):
      // Any live cell with fewer than two live neighbours dies (referred to as underpopulation).
      // Any live cell with more than three live neighbours dies (referred to as overpopulation).
      // Any live cell with two or three live neighbours lives, unchanged, to the next generation.
      // Any dead cell with exactly three live neighbours comes to life.
      // if (neighborCount === 3 || (neighborCount === 2 && occupiedTiles.has(key))) {

      // Simplest rule: just grow (designed for 4 neighbors):
      if (neighborCount >= 1 || occupiedTiles.has(key)) {
        if (!hitTestAllEntities(x, y).some(hit => !(hit.entity instanceof Block || hit.entity instanceof CellularAutomata))) {
          newOccupiedTiles.add(key)
        }
      }
    }
  }
  for (let i = entities.length - 1; i >= 0; i--) {
    const entity = entities[i]
    if (entity instanceof CellularAutomata) {
      const key = `${entity.x},${entity.y}`
      if (newOccupiedTiles.has(key)) {
        continue
      }
      if (!hitTestAllEntities(entity.x, entity.y).some(hit => !(hit.entity instanceof Block || hit.entity instanceof CellularAutomata))) {
        entities.splice(i, 1)
      }
    }
  }
  for (const key of newOccupiedTiles) {
    if (occupiedTiles.has(key)) {
      continue
    }
    const [x, y] = key.split(',').map(Number)
    const cell = new CellularAutomata()
    cell.x = x
    cell.y = y
    cell.layer = invertCollisionLayer(topLayer(hitTestAllEntities(x, y)))
    entities.push(cell)
  }
}

File: game/go/game_logic.go
package main

import "slices"

func AnalyzeMoveAbsolute(snake *Snake, targetTile Point, level *Level) Move {
	deltaGridX := int(targetTile.X - snake.Segments[0].X)
	deltaGridY := int(targetTile.Y - snake.Segments[0].Y)
	return AnalyzeMoveRelative(snake, deltaGridX, deltaGridY, level)
}

func AnalyzeMoveRelative(snake *Snake, deltaX, deltaY int, level *Level) Move {
	head := snake.Segments[0]
	x := head.X + deltaX
	y := head.Y + deltaY

	ignoreTailOfSnake := snake
	if snake.GrowOnNextMove {
		ignoreTailOfSnake = nil
	}
	hitsAhead := hitTestAllEntities(x, y, level, HitTestOptions{
		IgnoreTailOfSnake: ignoreTailOfSnake,
	})

	hitsAllAlong := []Hit{}
	for _, seg := range snake.Segments {
		hitsAllAlong = append(hitsAllAlong, hitTestAllEntities(seg.X, seg.Y, level, HitTestOptions{})...)
	}

	encumbered := false
	for _, hit := range hitsAllAlong {
		// if hit.Entity.Solid() &&
		// 	hit.Entity != snake &&
		// 	indexOfEntity(hit, level.Entity) > indexOfEntity(snake, level) &&
		// 	!(hit.EntityIsSnake() && snake.FusedSnakeIds.Contains(hit.EntityID())) {
		encumbered = hit.Entity != nil &&
			hit.Entity.IsSolid() &&
			hit.Entity != snake &&
			indexOfEntity(hit.Entity, level) > indexOfEntity(snake, level)
		if encumbered {
			break
		}
	}

	// Prevent moving backwards when two segments long
	// (When one segment long, you can plausibly move in any direction,
	// and when more than two segments long, a body segment will be in the way,
	// but when two segments, normally the tail is excluded from hit testing,
	// so you would be allowed to double back, but it feels weird to do a 180° turn.)
	// This also prevents an overlapped snake from doubling back on itself,
	// by moving into a tile occupied by a snake which is on top of this snake.
	// (But that is just a special case of the rule that you shouldn't
	// be able to move into a tile occupied by a snake which is on top of this snake.)
	movingBackwards := len(snake.Segments) > 1 &&
		deltaX == sign(snake.Segments[1].X-head.X) &&
		deltaY == sign(snake.Segments[1].Y-head.Y)

		// I think I will need to move to a system where the move is simulated and then checked for validity,
		// to avoid the complexity of adding exceptions to game state access, when answering hypotheticals.
		// This could also help with animating undo/redo, which currently replaces all the entities, resetting animation timers,
		// and for the level editor, where I'd like to check for collisions to show warnings,
		// (but allow the collisions to happen so that editing isn't its own puzzle.)
		// If moves are analyzed by checking for collisions within a whole game board,
		// it could share some code. Theoretically.

	// // Push objects
	// var entitiesToPush []entity.Entity
	// {
	// 	hit := FindFirstSolidEntity(hitsAhead)
	// 	// TODO: try pushing other snakes too (or save that for when splitting snakes; only push the tail / "dead" half)
	// 	// TODO: recursively push crates
	// 	if crateEntity, ok := hit.(*crate.Crate); ok {
	// 		newTile := Point{
	// 			X:      crateEntity.X + deltaX,
	// 			Y:      crateEntity.Y + deltaY,
	// 			Width:  crateEntity.Width,
	// 			Height: crateEntity.Height,
	// 		}
	// 		hitsAheadCrate := HitTestAllEntities(newTile.X, newTile.Y, HitTestOptions{IgnoreTailOfSnake: snake})
	// 		if WithinLevel(newTile) &&
	// 			LayersCollide(crateEntity.Layer, head.Layer) &&
	// 			!LayersCollide(TopLayer(hitsAheadCrate), crateEntity.Layer) {
	// 			entitiesToPush = append(entitiesToPush, crateEntity)
	// 			for _, h := range hitsAhead {
	// 				if _, ok := h.Entity.(*collectable.Collectable); ok {
	// 					entitiesToPush = append(entitiesToPush, h.Entity)
	// 					break
	// 				}
	// 			}
	// 		}
	// 	}
	// }

	// Ignore pushed objects as obstacles
	// hitsAhead = FilterHitsExcludingEntities(hitsAhead, entitiesToPush)

	return Move{
		// SnakeId:   s.ID,
		Snake: snake,
		Valid: (deltaX == 0 || deltaY == 0) &&
			(abs(deltaX) == 1 || abs(deltaY) == 1) &&
			withinLevel(Point{X: x, Y: y}, level) &&
			!movingBackwards &&
			!encumbered &&
			!layersCollide(topLayer(hitsAhead), snake.Layer),
		Encumbered:    encumbered,
		To:            Point{X: x, Y: y},
		Delta:         Point{X: deltaX, Y: deltaY},
		EntitiesThere: hitsToEntities(hitsAhead),
		// EntitiesToPush: entitiesToPush,
	}
}

func CanMove(snake *Snake, level *Level) bool {
	return AnalyzeMoveRelative(snake, 1, 0, level).Valid ||
		AnalyzeMoveRelative(snake, 0, 1, level).Valid ||
		AnalyzeMoveRelative(snake, -1, 0, level).Valid ||
		AnalyzeMoveRelative(snake, 0, -1, level).Valid
}

func TakeMove(m Move, level *Level) {
	s := m.Snake
	// originalTailPos := s.Segments[len(s.Segments)-1]

	// undoable() // handled externally
	// audio.PlaySound("move")

	if s.GrowOnNextMove {
		// Duplicate the tail segment
		tail := s.Segments[len(s.Segments)-1]
		s.Segments = append(s.Segments, tail)
		s.GrowOnNextMove = false
	}

	moveSnakeByHead(s, m.To)

	// s.Facing = m.Delta

	// Sort entities
	// Could probably just unconditionally sort the moved snake to the top.
	maxIndex := -1
	for _, e := range m.EntitiesThere {
		if e.IsSolid() {
			maxIndex = max(maxIndex, indexOfEntity(e, level))
		}
	}
	if maxIndex > -1 {
		thisIndex := indexOfEntity(s, level)
		for thisIndex < maxIndex {
			level.Entities[thisIndex], level.Entities[thisIndex+1] = level.Entities[thisIndex+1], level.Entities[thisIndex]
			thisIndex++
		}
	}

	// if len(m.EntitiesToPush) > 0 {
	// 	audio.PlaySound("pushCrate", audio.Options{
	// 		PlaybackRate: rand.Float64()*0.1 + 0.95,
	// 		Volume:       0.3,
	// 	})
	// }

	// for _, e := range m.EntitiesToPush {
	// 	TranslateEntity(e, m.Delta.X, m.Delta.Y)
	// 	game.RemoveEntity(e)
	// 	game.AddEntityToTop(e)
	// }

	// if len(m.EntitiesToPush) > 0 {
	// 	game.SortEntities()
	// }

	for _, e := range m.EntitiesThere {
		if c, ok := e.(*Food); ok { // TODO: collectables in general
			if layersCollide(c.Layer, s.Layer) { // TODO: && !entityInSlice(c, m.EntitiesToPush) {
				index := indexOfEntity(c, level)
				if index < 0 {
					continue // maybe
				}
				level.Entities = slices.Delete(level.Entities, index, index+1)
				// if _, isFood := c.(*food.Food); isFood {
				s.GrowOnNextMove = true
				// if !game.CheckLevelWon() {
				// 	audio.PlayMelodicSound("eat", s.NextMelodyIndex())
				// }
				// } else if _, isInverter := c.(*inverter.Inverter); isInverter {
				// 	InvertSnake(s)
				// }
			}
		}
	}

	// for id := range s.FusedSnakeIds {
	// 	if fs, ok := game.FindSnakeByID(id); ok {
	// 		DragSnake(fs, len(fs.Segments)-1, s.Segments[len(s.Segments)-1])
	// 	}
	// }

	// UpdateCellularAutomata()
	// s.AnimateMove(m, originalTailPos)
}

// // TODO: DRY, copied from function `drag` in level-editor.ts
// func DragSnake(dragging *Snake, index int, to Point) {
// 	segment := dragging.Segments[index]
// 	if segment.X != to.X || segment.Y != to.Y {
// 		from := Point{X: segment.X, Y: segment.Y}
// 		points := LineNoDiagonals(from, to)
// 		for _, point := range points[1:] {
// 			for i := len(dragging.Segments) - 1; i > index; i-- {
// 				lead(dragging.Segments[i-1], dragging.Segments[i])
// 			}
// 			for i := 0; i < index; i++ {
// 				lead(dragging.Segments[i+1], dragging.Segments[i])
// 			}
// 			segment.X = point.X
// 			segment.Y = point.Y
// 			// if snake.DEBUG_SNAKE_DRAGGING {
// 			//   draw()
// 			// }
// 		}
// 	}
// }

// // TODO: DRY, copied from function `lead` in level-editor.ts
// func lead(leader, follower *snake.Segment) {
// 	follower.X = leader.X
// 	follower.Y = leader.Y
// 	follower.Width = leader.Width
// 	follower.Height = leader.Height
// 	// if snake.DEBUG_SNAKE_DRAGGING {
// 	//   draw()
// 	// }
// }

/*
func InvertSnake(s *Snake) {
  // See also: `invert` function in level-editor.ts
	handledEntities := make(map[entity.Entity]bool)
	handledPositions := make(map[string]bool)

	var handleEntity func(entity.Entity)
	var handlePosition func(int, int)

	handleEntity = func(e entity.Entity) {
		if handledEntities[e] {
			return
		}
		handledEntities[e] = true
		if re, ok := e.(*rectangularentity.RectangularEntity); ok {
			re.Layer = InvertCollisionLayer(re.Layer)
		} else if snk, ok := e.(*Snake); ok {
			for _, seg := range snk.Segments {
				handlePosition(seg.X, seg.Y)
				seg.Layer = InvertCollisionLayer(seg.Layer)
			}
		}
	}

	handlePosition = func(x, y int) {
		key := fmt.Sprintf("%d,%d", x, y)
		if handledPositions[key] {
			return
		}
		handledPositions[key] = true
		for _, hit := range HitTestAllEntities(x, y) {
			handleEntity(hit.Entity)
		}
	}

	handleEntity(s)

	for pos := range handledPositions {
		var x, y int
		fmt.Sscanf(pos, "%d,%d", &x, &y)
		hits := HitTestAllEntities(x, y)
		if !AnyBlock(hits) {
			block := &block.Block{
				X:     x,
				Y:     y,
				Layer: types.CollisionLayerWhite,
			}
			game.AddEntityToBottom(block)
		}
	}
}

func UpdateCellularAutomata() {
	occupied := map[string]*cellularautomata.CellularAutomata{}
	for _, e := range game.Entities {
		if ca, ok := e.(*cellularautomata.CellularAutomata); ok {
			key := fmt.Sprintf("%d,%d", ca.X, ca.Y)
			occupied[key] = ca
		}
	}

	newOccupied := map[string]bool{}
	for y := 0; y < game.LevelInfo.Height; y++ {
		for x := 0; x < game.LevelInfo.Width; x++ {
			key := fmt.Sprintf("%d,%d", x, y)
			neighbors := []string{
				fmt.Sprintf("%d,%d", x, y-1),
				fmt.Sprintf("%d,%d", x+1, y),
				fmt.Sprintf("%d,%d", x, y+1),
				fmt.Sprintf("%d,%d", x-1, y),
			}

			count := 0
			for _, n := range neighbors {
				if occupied[n] != nil {
					layer := TopLayer(FilterHitsExcludingType(HitTestAllEntities(x, y), &cellularautomata.CellularAutomata{}))
					if occupied[n].Layer == InvertCollisionLayer(layer) {
						count++
					}
				}
			}

			if count >= 1 || occupied[key] != nil {
				if AllEntitiesAreBlocksOrCA(HitTestAllEntities(x, y)) {
					newOccupied[key] = true
				}
			}
		}
	}

	for i := len(game.Entities) - 1; i >= 0; i-- {
		if ca, ok := game.Entities[i].(*cellularautomata.CellularAutomata); ok {
			key := fmt.Sprintf("%d,%d", ca.X, ca.Y)
			if !newOccupied[key] && AllEntitiesAreBlocksOrCA(HitTestAllEntities(ca.X, ca.Y)) {
				game.Entities = append(game.Entities[:i], game.Entities[i+1:]...)
			}
		}
	}

	for key := range newOccupied {
		if occupied[key] == nil {
			var x, y int
			fmt.Sscanf(key, "%d,%d", &x, &y)
			cell := &cellularautomata.CellularAutomata{
				X:     x,
				Y:     y,
				Layer: InvertCollisionLayer(TopLayer(HitTestAllEntities(x, y))),
			}
			game.Entities = append(game.Entities, cell)
		}
	}
}
*/
